internal/component/unit.go

```diff
@@ -188,3 +216,125 @@ func (u *Unit) Dependents() Components {
 
 	return u.dependents
 }
+
+// String renders this unit as a human-readable string for debugging.
+func (u *Unit) String() string {
+	// Snapshot values under read lock to avoid data races
+	u.rLock()
```

Comment:
```suggestion
	u.rLock()
	defer u.rUnlock()
```

use defer to keep the cleanup code close to the setup code